hbase-server/src/test/java/org/apache/hadoop/hbase/replication/regionserver/TestReplicationSourceWithoutReplicationZnodes.java

```diff
@@ -0,0 +1,234 @@
+    manager = replication.getReplicationManager();
+
+    manager.addSource(slaveId);
+
+    htd = new HTableDescriptor(test);
+    HColumnDescriptor col = new HColumnDescriptor(f1);
+    col.setScope(HConstants.REPLICATION_SCOPE_GLOBAL);
+    htd.addFamily(col);
+    col = new HColumnDescriptor(f2);
+    col.setScope(HConstants.REPLICATION_SCOPE_LOCAL);
+    htd.addFamily(col);
+
+    hri = new HRegionInfo(htd.getTableName(), r1, r2);
+  }
+
+  @AfterClass
+  public static void tearDownAfterClass() throws Exception {
+    try {
+      manager.join();
+    } catch (RuntimeException re) {
+      if (re.getMessage().equals(fakeExceptionMessage)) {
+        LOG.info("It is fine");
+      }
+    }
+    utility.shutdownMiniCluster();
+  }
+
+  @Rule
+  public TestName testName = new TestName();
+
+  private void cleanLogDir() throws IOException {
+    fs.delete(logDir, true);
+    fs.delete(oldLogDir, true);
+  }
+
+  @Before
+  public void setUp() throws Exception {
+    LOG.info("Start " + testName.getMethodName());
+    cleanLogDir();
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    LOG.info("End " + testName.getMethodName());
+    cleanLogDir();
+  }
+
+  /**
+   * When the peer is removed, hbase remove the peer znodes and there is zk watcher
+   * which terminates the replication sources. In case of zk watcher not getting invoked
+   * or a race condition between source deleting the log znode and zk watcher
+   * terminating the source, we might get the NoNode exception. In that case, the right
+   * thing is to terminate the replication source.
+   * @throws Exception
+   */
+  @Test
+  public void testReplicationSourceRunningWithoutPeerZnodes() throws Exception {
+    MultiVersionConcurrencyControl mvcc = new MultiVersionConcurrencyControl();
+    KeyValue kv = new KeyValue(r1, f1, r1);
+    WALEdit edit = new WALEdit();
+    edit.add(kv);
+
+    List<WALActionsListener> listeners = new ArrayList<>();
+    listeners.add(replication);
+    final WALFactory wals = new WALFactory(utility.getConfiguration(), listeners,
+      URLEncoder.encode("regionserver:60020", "UTF8"));
+    final WAL wal = wals.getWAL(hri.getEncodedNameAsBytes(), hri.getTable().getNamespace());
+    manager.init();
+
+    final long txid = wal.append(htd, hri,
+      new WALKey(hri.getEncodedNameAsBytes(), test, System.currentTimeMillis(), mvcc),
+      edit, true);
+    wal.sync(txid);
+
+
+    wal.rollWriter();
+    ZKUtil.deleteNodeRecursively(zkw, "/hbase/replication/peers/1");
+    ZKUtil.deleteNodeRecursively(zkw, "/hbase/replication/rs/"+ server.getServerName() + "/1");
+
+    ReplicationException exceptionThrown = null;
+    try {
+      manager.logPositionAndCleanOldLogs(manager.getSources().get(0).getCurrentPath(), "1", 0, false,
+        false);
+    } catch (ReplicationException e) {
+      exceptionThrown = e;
+
+    }
+
+    Assert.assertTrue(exceptionThrown instanceof ReplicationSourceWithoutPeerException);
```

Comment:
Would like to see one test where we didn't encounter `KeeperException.NoNodeException` in `ReplicationQueuesZKImpl#removeLog`   but some other `KeeperException` and verify that `RS#terminate` was called.

Reply:
@shahrs87 I think in that case, the RS will abort.

Reply:
We can mock Abortable and verify whether abort is called or nor ? Would that work ?

Reply:
@shahrs87 thanks for your feedback, I have added the test for source termination on replication exception for no peer.